crypto/kmac/kmac.c

```diff
@@ -0,0 +1,460 @@
+ */
+#define KMAC_MAX_KEY_ENCODED (KMAC_MAX_BLOCKSIZE * 2)
+
+/* Fixed value of encode_string("KMAC") */
+static const unsigned char kmac_string[] = {
+    0x01, 0x20, 0x4B, 0x4D, 0x41, 0x43
+};
+
+/* If this flag is set then XOF mode is used (i.e right_encode(0)) */
+#define KMAC_FLAG_XOF_MODE          1
+
+/* typedef EVP_MAC_IMPL */
+struct evp_mac_impl_st {
+    EVP_MD_CTX *ctx;
+    size_t out_len;
+    int key_len;
+    int custom_len;
+    int flags;
+    /* key and custom are stored in encoded form */
+    unsigned char key[KMAC_MAX_KEY_ENCODED];
+    unsigned char custom[KMAC_MAX_CUSTOM_ENCODED];
+};
+
+static int encode_string(unsigned char *out, int *out_len,
+                         const unsigned char *in, int in_len);
+static int right_encode(unsigned char *out, int *out_len, size_t bits);
+static int bytepad(unsigned char *out, int *out_len,
+                   const unsigned char *in1, int in1_len,
+                   const unsigned char *in2, int in2_len,
+                   int w);
+static int kmac_bytepad_encode_key(unsigned char *out, int *out_len,
+                                   const unsigned char *in, int in_len,
+                                   int w);
+static int kmac_ctrl_str(EVP_MAC_IMPL *kctx, const char *type,
+                         const char *value);
+
+
+static void kmac_free(EVP_MAC_IMPL *kctx)
+{
+    if (kctx != NULL) {
+        EVP_MD_CTX_free(kctx->ctx);
+        OPENSSL_cleanse(kctx->key, kctx->key_len);
+        OPENSSL_cleanse(kctx->custom, kctx->custom_len);
+        OPENSSL_free(kctx);
+    }
+}
+
+static EVP_MAC_IMPL *kmac_new(void)
+{
+    EVP_MAC_IMPL *kctx;
+
+    if ((kctx = OPENSSL_zalloc(sizeof(*kctx))) == NULL
+        || (kctx->ctx = EVP_MD_CTX_new()) == NULL) {
+        kmac_free(kctx);
+        return NULL;
+    }
+    return kctx;
+}
+
+static int kmac_copy(EVP_MAC_IMPL *gdst, EVP_MAC_IMPL *gsrc)
+{
+    gdst->out_len = gsrc->out_len;
+    gdst->key_len = gsrc->key_len;
+    gdst->custom_len = gsrc->custom_len;
+    gdst->flags = gsrc->flags;
+    memcpy(gdst->key, gsrc->key, gsrc->key_len);
+    memcpy(gdst->custom, gsrc->custom, gdst->custom_len);
+
+    return EVP_MD_CTX_copy(gdst->ctx, gsrc->ctx);
+}
+
+/*
+ * The init() assumes that any ctrl methods are set beforehand for
+ * md, key and custom. Setting the fields afterwards will have no
+ * effect on the output mac.
+ */
+static int kmac_init(EVP_MAC_IMPL *kctx)
+{
+    EVP_MD_CTX *ctx = kctx->ctx;
+    unsigned char out[KMAC_MAX_BLOCKSIZE];
+    int out_len, block_len;
+
+    /* Only check for key here (setting the key is dependent on the digest) */
+    if (kctx->key_len == 0) {
+        EVPerr(EVP_F_KMAC_INIT, EVP_R_NO_KEY_SET);
+        return 0;
+    }
+    block_len = EVP_MD_CTX_block_size(ctx);
+
+    /* Set default custom string if it is not already set */
+    if (kctx->custom_len == 0)
+        (void)kmac_ctrl_str(kctx, "custom", "");
+
+    return bytepad(out, &out_len, kmac_string, sizeof(kmac_string),
```

Comment:
Do we support a sequence like:
```
init()
init()
update()
final()
```

where the effects of the first `init` are ignored?

I don't see why we should and if that assumption is correct, the code is here if fine.

Reply:
Hmm this should proabably be made to work.. So I will cache md and not do the digestInit to the kmac_init().